component/dialer/sockopt_windows.go

```diff
@@ -0,0 +1,67 @@
+package dialer
+
+import (
+	"encoding/binary"
+	"net"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+func setSocketOptions(network, address string, c syscall.RawConn, opts *Options) (err error) {
+	if opts == nil || !isTCPSocket(network) && !isUDPSocket(network) {
+		return
+	}
+
+	var innerErr error
+	err = c.Control(func(fd uintptr) {
+		host, _, _ := net.SplitHostPort(address)
+		if ip := net.ParseIP(host); ip != nil && !ip.IsGlobalUnicast() {
+			return
+		}
+
+		if opts.InterfaceIndex == 0 && opts.InterfaceName != "" {
+			if iface, err := net.InterfaceByName(opts.InterfaceName); err == nil {
+				opts.InterfaceIndex = iface.Index
+			}
+		}
+
+		if opts.InterfaceIndex != 0 {
+			switch network {
+			case "tcp4", "udp4":
+				innerErr = bindSocketToInterface4(windows.Handle(fd), uint32(opts.InterfaceIndex))
+			case "tcp6", "udp6":
+				innerErr = bindSocketToInterface6(windows.Handle(fd), uint32(opts.InterfaceIndex))
+				if network == "udp6" {
```

Comment:
I have two doubts here:

1. What happens under UDP6 circumstance since we bind to the interface 6 first and then bind to the interface 4? (double bind?)
2. Have you try to convert the ipv6 address to ipv4 first, and see if the address can be bind to ipv4 directly?

Like this:
```go
case "udp6":
  if addr.To4() {
    bindToV4()
  else {
    bindToV6()
  }
```

Reply:
1. I'm not sure what happens. But via my testing, I belive that it would bind the fd to both IPv4 and IPv6 proto under UDP6 circumstance. 
2. I've tried. But that is impossible under UDP circumstance, there are two reason:
    * the address param always will be ":0" in most common case, since the address param will be passed by empty string when executing `dialer.ListenPacketWithOptions` func
    * if we don't pass empty string for address param when executing `dialer.ListenPacketWithOptions` func:
         * passed IPv4 address: the network param will be `udp4`, so the code logic will not into `udp6` case. 
         * passed IPv6 address: the network param will be `udp6`, in this case we don't need to bind interface4. (I've updated the code to cover this case.)

Reply:
Not the address variable. There’s a `ip` variable can be used to determine v4 or v6.

Reply:
`ip` is parsed from `address`, right?  So the result is the same as my said above.

Reply:
Oh right, I misunderstood before.

Reply:
And I've done a mini update: add `ip.To16() == nil `
```go
if network == "udp6" && ip.To16() == nil {
	innerErr = bindSocketToInterface4(windows.Handle(fd), uint32(opts.InterfaceIndex))
}
```